src/Bicep.Core/Syntax/ISyntaxHierarchy.cs

```diff
@@ -0,0 +1,66 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+
+namespace Bicep.Core.Syntax
+{
+    public interface ISyntaxHierarchy
+    {
+        /// <summary>
+        /// Gets the parent of the specified node. Returns null for root nodes. Throws an exception for nodes that have not been indexed.
+        /// </summary>
+        /// <param name="node">The node</param>
+        SyntaxBase? GetParent(SyntaxBase node);
+
+        /// <summary>
+        /// Gets all ancestor nodes assignable to <typeparamref name="TSyntax" /> in descending order
+        /// from the top of the tree.
+        /// </summary>
+        /// <param name="syntax">The syntax node.</param>
+        /// <typeparam name="TSyntax">The type of node to query.</typeparam>
+        /// <returns>The list of ancestors.</returns>
+        public ImmutableArray<TSyntax> GetAllAncestors<TSyntax>(SyntaxBase syntax) where TSyntax : SyntaxBase
```

Comment:
>GetAllAncestors [](start = 39, length = 15)

Default interface implementations ftw! ([https://devblogs.microsoft.com/dotnet/default-implementations-in-interfaces/](https://devblogs.microsoft.com/dotnet/default-implementations-in-interfaces/))

Reply:
FYI the sharp edge with DIMs is that they are **effectively** similar to explicit interface implementations.

If you have an interface `IDog` with a DIM `GetSound() => "woof"`. Then the class `Chihuahua` doesn't have `GetSound()`.

```C#
var taco = new Chihuahua();
taco.GetSound(); // does not compile

((IDog)taco).GetSound(); // happy compiler
```

They are good when your consumers use the interface, and not good when you consumers use the class.